velox/common/memory/MappedMemory.h

```diff
@@ -26,12 +26,88 @@
 #include <gflags/gflags.h>
 #include "velox/common/base/Exceptions.h"
 #include "velox/common/memory/MemoryUsageTracker.h"
+#include "velox/common/time/Timer.h"
 
 DECLARE_bool(velox_use_malloc);
 DECLARE_int32(velox_memory_pool_mb);
 
 namespace facebook::velox::memory {
 
+struct SizeStats {
+  int32_t size{0};
+  int64_t allocClocks{0};
+  int64_t freeClocks{0};
+  int64_t numAlloc{0};
+  int64_t cumBytes{0};
+
+  SizeStats operator-(const SizeStats& other) const {
+    SizeStats result;
+    result.size = size;
+    result.allocClocks = allocClocks - other.allocClocks;
+    result.allocClocks = freeClocks - other.freeClocks;
+    result.numAlloc = numAlloc - other.numAlloc;
+    result.cumBytes = cumBytes - other.cumBytes;
+    return result;
+  }
+
+  uint64_t clocks() const {
+    return allocClocks + freeClocks;
+  }
+};
+
+struct Stats {
+  static constexpr int32_t kNumSizes = 20;
+  Stats() {
+    for (auto i = 0; i < sizes.size(); ++i) {
+      sizes[i].size = 1 << i;
+    }
+  }
+
+  Stats operator-(const Stats& stats) const;
+
+  template <typename Op>
+  void recordAlloc(int64_t bytes, int32_t count, Op op) {
+    uint64_t clocks{0};
+    auto index = sizeIndex(bytes);
+    {
+      velox::ClockTimer timer(clocks);
+      op();
+    }
+    sizes[index].numAlloc += count;
+    sizes[index].cumBytes += bytes * count;
+    sizes[index].allocClocks += clocks;
+    if (0) {
+      auto trace = std::make_unique<process::StackTrace>();
+      LOG(INFO) << "XLARGE: Alloc size " << (1 << index) << ": "
+                << trace->toString();
+    }
+  }
+
+  template <typename Op>
+  void recordFree(int64_t bytes, Op op) {
+    uint64_t clocks = 0;
+    auto index = sizeIndex(bytes);
+    {
+      ClockTimer timer(clocks);
+      op();
+    }
+    sizes[index].freeClocks += clocks;
+  }
+
+  std::string toString() const;
+
+  int32_t sizeIndex(int64_t size) {
```

Comment:
I wonder if we don't have this calculation somewhere else already and whether it would be possible to duplicate. I'm imagining there must be code that receives a request to allocate N bytes and figures our which size class to use for that.